enderio-machines/src/main/java/com/enderio/machines/common/blockentity/task/PoweredCraftingTask.java

```diff
@@ -0,0 +1,285 @@
+            return;
+        }
+
+        // If we haven't done so already, consume inputs for the recipe.
+        if (!collectedInputs) {
+            // Consume inputs for the recipe.
+            takeInputs(recipe);
+            collectedInputs = true;
+
+            // Store the recipe energy cost.
+            // This is run afterwards as it allows container context changes after takeInputs()
+            energyCost = recipe.getEnergyCost(container);
+        }
+
+        // Try to consume as much energy as possible to finish the craft.
+        if (energyConsumed <= energyCost) {
+            energyConsumed += consumeEnergy(energyCost - energyConsumed);
+        }
+
+        // If the recipe has been crafted, attempt to put it into storage
+        if (energyConsumed >= energyCost) {
+            // Attempt to complete the craft
+            if (placeOutputs(outputs, false)) {
+                // The receiver was able to take the outputs, task complete.
+                complete = true;
+            }
+        }
+    }
+
+    @Override
+    public float getProgress() {
+        if (recipe == null)
+            return 0.0f;
+        return energyConsumed / (float) recipe.getEnergyCost(container);
+    }
+
+    @Override
+    public boolean isComplete() {
+        return complete;
+    }
+
+    // region Serialization
+
+    @Override
+    public CompoundTag serializeNBT() {
+        CompoundTag tag = new CompoundTag();
+        tag.put("Recipe", serializeRecipe(new CompoundTag(), recipe));
+        tag.putInt("EnergyConsumed", energyConsumed);
+        tag.putInt("EnergyCost", energyConsumed);
+        tag.putBoolean("CollectedInputs", collectedInputs);
+        tag.putBoolean("Complete", complete);
+
+        tag.putBoolean("DeterminedOutputs", determinedOutputs);
+        if (determinedOutputs) {
+            ListTag outputsNbt = new ListTag();
+            for (OutputStack stack : outputs) {
+                outputsNbt.add(stack.serializeNBT());
+            }
+            tag.put("Outputs", outputsNbt);
+        }
+
+        return tag;
+    }
+
+    @Override
+    public void deserializeNBT(CompoundTag nbt) {
+        recipe = loadRecipe(nbt.getCompound("Recipe"));
+        energyConsumed = nbt.getInt("EnergyConsumed");
+        energyCost = nbt.getInt("EnergyCost");
+        collectedInputs = nbt.getBoolean("CollectedInputs");
+        complete = nbt.getBoolean("Complete");
+
+        determinedOutputs = nbt.getBoolean("DeterminedOutputs");
+        if (determinedOutputs) {
+            ListTag outputsNbt = nbt.getList("Outputs", Tag.TAG_COMPOUND);
+            outputs = new ArrayList<>();
+            for (Tag tag : outputsNbt) {
+                outputs.add(OutputStack.fromNBT((CompoundTag) tag));
+            }
+        }
+    }
+
+    /**
+     * Serialize the recipe.
+     * Override this if something else needs to be saved.
+     */
+    protected CompoundTag serializeRecipe(CompoundTag tag, R recipe) {
```

Comment:
Consider dropping this custom serializer business and only support the custom loader. This feature was superseded by container contexts.